File: app/src/main/java/su/linka/linkapaperboard/GridController.java
package su.linka.linkapaperboard;

import android.text.TextUtils;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.GridView;

import java.util.Arrays;

public class GridController implements AdapterView.OnItemClickListener {

    private static GridController instance;
    private static View context;
    private final GridView grid;
    private int page = 0;
    private String text;
    private int gridSize;
    private OnKeyListener onKeyListener;
    private SlideButtonsController sbc;


    public static GridController getInstance() {
        if (instance == null || (context != MainActivity.context.getWindow().getDecorView())) {
            instance = new GridController(MainActivity.context.getWindow().getDecorView());
            instance.setOnKeyListener(new OnKeyListener() {
                @Override
                public void press(String k) {
                    TextViewController.getInstance().write(k);
                }
            });
        }
        return instance;
    }

    GridController(View context) {

        GridController.context = context;
        sbc = new SlideButtonsController(context);
        text = "␣" + GridController.context.getResources().getString(R.string.alphabet);
        grid = (context).findViewById(R.id.main_grid);
        grid.setOnItemClickListener(this);

        Button leftButton = context.findViewById(R.id.left_btn);
        Button rightButton = context.findViewById(R.id.right_btn);
        final GridController gc = this;
        leftButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                gc.previosPage();
            }
        });

        rightButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                gc.nextPage();
            }
        });


        gridSize = Cookie.getInstance(context.getContext()).getGridSize();

        draw();

    }

    public void draw() {

        int size = gridSize * gridSize;

        grid.setAdapter(new GridItemController(context.getContext(), R.layout.grid_button, getPageArray(page)));
        sbc.setTextForLeftBtn(TextUtils.join(", ", getPageArray(page == 0 ? (text.length() / size) : page - 1)));
        sbc.setTextForRightBtn((TextUtils.join(", ", getPageArray(page == (text.length() / size) ? 0 : page + 1))));
    }

    private String[] getPageArray(int page) {
        int size = gridSize;
        grid.setNumColumns(size);
        size *= size;


        String localtext = text.substring(page * size, Math.min(text.length(), size * page + size));
        return Arrays.copyOfRange(localtext.split(""), 1, localtext.length() + 1);
    }

    public void previosPage() {
        if (page == 0) {
            page = text.length() / (gridSize * gridSize);
        } else {
            page--;
        }
        draw();
    }

    public void nextPage() {
        int size = gridSize * gridSize;
        if (text.length() < size * page + size) {
            page = 0;
        } else {
            page++;
        }
        draw();
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        onKeyListener.press((text.charAt(gridSize * gridSize * page + i) + "").replace("␣", " "));
    }

    public int getSize() {
        return gridSize;
    }

    public float getGridHeight() {
        return grid.getMeasuredHeight();
    }

    public void setGridSize(int gridSize) {
        this.gridSize = gridSize;
        Cookie.getInstance().setGridSize(gridSize);
        page = 0;
        draw();
    }

    public void setText(String text) {
        this.text = text;
        draw();
    }

    public void setOnKeyListener(OnKeyListener onKeyListener) {
        this.onKeyListener = onKeyListener;
    }

    public static abstract class OnKeyListener {
        public abstract void press(String k);
    }
}

File: app/src/main/java/su/linka/linkapaperboard/GridController.kt
package su.linka.linkapaperboard

import android.text.TextUtils
import android.view.View
import android.widget.AdapterView
import android.widget.GridView
import kotlin.math.min

class GridController(
    private val contextView: View,
    private val preferences: Cookie,
    private val uppercaseProvider: () -> Boolean,
    private val onKey: (String) -> Unit
) : AdapterView.OnItemClickListener, View.OnClickListener {

    private val grid: GridView = contextView.findViewById(R.id.main_grid)
    private var page = 0
    private var text: String = "␣" + contextView.resources.getString(R.string.alphabet)
    private var gridSize: Int = preferences.gridSize
    private val slideButtonsController = SlideButtonsController(contextView)

    init {
        grid.onItemClickListener = this
        contextView.findViewById<View>(R.id.left_btn).setOnClickListener(this)
        contextView.findViewById<View>(R.id.right_btn).setOnClickListener(this)
        draw()
    }

    fun draw() {
        val size = gridSize * gridSize
        val cellSizeProvider = { gridHeight / (gridSize + 2) }
        grid.adapter = GridItemController(
            contextView.context,
            R.layout.grid_button,
            getPageArray(page),
            cellSizeProvider,
            uppercaseProvider
        )
        slideButtonsController.setTextForLeftBtn(
            TextUtils.join(", ", getPageArray(if (page == 0) text.length / size else page - 1))
        )
        slideButtonsController.setTextForRightBtn(
            TextUtils.join(", ", getPageArray(if (page == text.length / size) 0 else page + 1))
        )
    }

    private fun getPageArray(page: Int): Array<String> {
        var size = gridSize
        grid.numColumns = size
        size *= size

        val start = page * size
        val end = min(text.length, size * page + size)
        val localText = text.substring(start, end)
        return localText.map { it.toString() }.toTypedArray()
    }

    fun previosPage() {
        page = if (page == 0) {
            text.length / (gridSize * gridSize)
        } else {
            page - 1
        }
        draw()
    }

    fun nextPage() {
        val size = gridSize * gridSize
        page = if (text.length < size * page + size) {
            0
        } else {
            page + 1
        }
        draw()
    }

    override fun onItemClick(adapterView: AdapterView<*>?, view: View?, position: Int, id: Long) {
        val index = gridSize * gridSize * page + position
        onKey(text[index].toString().replace("␣", " "))
    }

    val size: Int
        get() = gridSize

    val gridHeight: Float
        get() = grid.measuredHeight.toFloat()

    fun setGridSize(gridSize: Int) {
        this.gridSize = gridSize
        preferences.setGridSize(gridSize)
        page = 0
        draw()
    }

    fun setText(text: String) {
        this.text = text
        draw()
    }

    fun refreshUppercase() {
        draw()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.left_btn -> previosPage()
            R.id.right_btn -> nextPage()
        }
    }
}
